Throw error message when a user with email address as uid exists

sharidas opened the pull request:
When admin creates user we should check the users
with username as email address exists.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Lets consider a scenario where a user with username as email address exist ( for instance a guest user). And admin tries to create another user say `user1` with the email address of guest user. The user should not be created. And an error message should be displayed. This PR addresses the same. I have made the error message same as for:
a) when the user exists and admin tries to create the same user.
b) The user names differ, but one of the username is same as the email address of new user to be created.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
A check to forbid the user creation when a user is created with same email address of a guest user.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Create a guest user with a valid email address
- From the webUI create another user with different uid and with same email address of guest.
- Error is thrown in the webUI.
- Also this has been tested using occ command and provisioning api.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

sharidas commented:
While looking into the UsersController.php there is a room for improvement to use UserCreateService. I  was thinking to make it as a separate PR/task. Let me know if that's ok or not.

jvillafanez commented:
I'm not sure if this is a good idea.

Users have an uid and an email (among other things). Whether the guests app uses the email as uid is irrelevant because for core is just another uid, what is important is that the guests app also sets the email as email.
Now, the question is, should 2 different users have the same email? I don't remember if there is already a DB constraint preventing multiple emails. In any case, I expect core to be handling this case already, no matter if the same email is duplicated or not.

sharidas commented:
>  In any case, I expect core to be handling this case already, no matter if the same email is duplicated or not.

If I understand the comment correctly you mean to say the core might be handling case of same email address already. But as per my observation the core does not. The core allows to have same email address for different users. In fact I wonder why can't we allow multiple users with same email address.  Or say what would cause wrong if 2 different users have same email address. Though I know that allowing multiple users having same email address would contradict to this PR ( that's fine ).

jvillafanez commented:
If we want to prevent duplicate emails, a DB constraint is the way to go. The fact that there is a DB index which allows duplicated values in the email means that core is fine allowing duplicated values for whatever reason.
The fix, if we want it, is to include that constraint in the DB ensuring a proper migration in case duplicates already exist in the DB.

> If I understand the comment correctly you mean to say the core might be handling case of same email address already. But as per my observation the core does not.

The solution, as said, is to put a constraint in the DB. If core hasn't made that change yet, I guess core is fine having duplicated emails for whatever reason.
My point here is that there is no bug if core allows it and handles the situation correctly.

> In fact I wonder why can't we allow multiple users with same email address. Or say what would cause wrong if 2 different users have same email address

Emails are intended to be personal and private. Admins use email to comunicate with the users. This can involve password resets, activity traces, and other personal information that you don't want anyone to see. Obviously, if the user uses public or shared email accounts, it's his responsability

sharidas commented:
> The fix, if we want it, is to include that constraint in the DB ensuring a proper migration in case duplicates already exist in the DB.

Totally agree with you. The only problem is when we touch the DB, we need to have migrations for the same.

> My point here is that there is no bug if core allows it and handles the situation correctly.

Agreed.

sharidas committed:
Make email address unique

The email address associated with username
should be unique.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Modified the PR with following changes:
- [x] Added migration to remove the duplicate email address with `NULL` and then mark the `email` column of accounts to unique. This would prevent duplicate email addresses. The draw back I see here is that the migration is setting duplicate email addresses to NULL. Meaning say there are 2 email addresses found with `foo@bar.com` then both are set to `NULL`. But then the admin/users have to take care to set the email address back. 
- [x] The warning message is thrown to the UI saying duplicate email address found.

GrendelWWU commented:
I guess, this is not an good idea. Different users may have the same mail-address for testing purposes.
